gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/impl/jdbc/authentication/LDAPAuthenticationSchemeImpl.java

```diff
@@ -724,198 +738,343 @@ private String getSearchAuthPwd() throws Exception {
-         */
-        private void resolveDNForGroup(DirContext ctx, String baseDN,
-            String searchFilter, String[] searchAttributes, boolean topLevel,
-            String group, List<String> groupMembers) throws NamingException {
 
-          NamingEnumeration<SearchResult> results;
-          // Construct Search Filter
-          SearchControls ctls = new SearchControls();
-          if (topLevel) {
-            // Setup an LDAP subtree search scope
-            ctls.setSearchScope(SearchControls.SUBTREE_SCOPE);
-          } else {
-            // Setup an LDAP single object search scope
-            ctls.setSearchScope(SearchControls.OBJECT_SCOPE);
+  /**
+   * Search for the LDAP group's DN in the LDAP server. LDAP server bind
+   * may or not be anonymous.
+   * <p>
+   * If the admin does not want us to do anonymous bind/search, then we
+   * must have been given principal/credentials in order to successfully
+   * bind to perform the user's DN search.
+   *
+   * @throws NamingException on failure to retrieve the LDAP group's base DN
+   */
+  public Set<String> getLDAPGroupMembers(String ldapGroup)
+    throws Exception {
+    ArrayList<String> groupMembers = new ArrayList<String>();
+    // Return all the unique members collected for the group
+    this.recurseGroups(ldapGroup, groupMembers, null);
+    return new UnifiedSet<>(groupMembers);
+  }
+
+
+
+
+
+  /**
+   * Search for the all the Groups the user is member of, the LDAP server. LDAP server bind
+   * may or not be anonymous.
+   * <p>
+   * If the admin does not want us to do anonymous bind/search, then we
+   * must have been given principal/credentials in order to successfully
+   * bind to perform the user's DN search.
+   *
+   * @throws NamingException on failure to retrieve the LDAP group's base DN
+   */
+  public Set<String> getLdapGroupsOfUser(String uid)
+    throws Exception {
+
+    DirContext ctx = getDirContext(uid);
+    String attributesNeeded[] = Arrays.copyOf(attrDN, attrDN.length + 1);
+    attributesNeeded[attributesNeeded.length - 1] = memberOfAttr;
+    NamingEnumeration results = getUserInformation(uid, ctx, attributesNeeded);
+    // If we did not find anything then login failed
+    if (results == null || !results.hasMore())
+      throw new NameNotFoundException();
+
+    SearchResult result = (SearchResult) results.next();
+    ArrayList<String> ldapGroups = new ArrayList<String>();
+    Attribute memberOfAttr = result.getAttributes().get("memberOf");
+    if (memberOfAttr != null) {
+      NamingEnumeration ne = memberOfAttr.getAll();
+      while (ne.hasMore()) {
+        String group = (String) ne.next();
+        LdapName groupDn = new LdapName(group);
+        Iterator<Rdn> iter = groupDn.getRdns().iterator();
+
+        while (iter.hasNext()) {
+          Rdn rdn = iter.next();
+          if (rdn.getType().equalsIgnoreCase("cn")) {
+            String cn = rdn.getValue().toString();
```

Comment:
Why should this be restricted to CN?

Reply:
I suppose UID needs to be considered too. Will add it.

Reply:
Actually , would it be UID? We are identifying the groups. I thought groups can be identified only using DN. Do groups also have a uid  or something like gid?